### PR TITLE
remove bouncy castle dependency from edgeHub

### PR DESCRIPTION
**Pull request by vipeller**
A quick solution to see how bouncy castle could be removed. <del>What missing are:
<del>- error handling
<del>- Import apparently does not like PKCS-8

Now it contains the full solution

### CODE REVIEW

**Review by varunpuranik — COMMENTED** [review]
Changes look good.. 
- Did you try removing the dependency as well? 
- Did you get a chance to test it as well? Does it work?
- Is this all that we need to do?

**Comment by vipeller** [comment]
I have not removed the dependency to the package yet, just the 'using'-s.
I started it from visual studio with an rsa cerrtificate, that worked.
I need to test it with EC, and also with pkcs8 keys (that I realized that those would not work with the ImportRSAPrivateKey() method, but from Martin's code it turned out that there is a separate ImportPkcs8PrivateKey() or similar call to read those)
The unit tests fail because those use pkcs8 certs. Once I add to import to pkcs8, they should start working.
The code will be a little-bit more complex with the pkcs8 part and with the error handling, but not so much.
I think it is about 1-2 to finish with the code, then I need to test it with 3 more key types (EC pkcs1/pkcs8 and rsa pkcs8 (I've tested rsa pkcs1 because that is what I have with my setup)

I've put together this code quickly only just to see the magnitude of the work, but it was easier than I thought.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 1]
```diff
@@ -9,18 +9,21 @@ namespace Microsoft.Azure.Devices.Edge.Util
...
-    using Org.BouncyCastle.OpenSsl;
-    using Org.BouncyCastle.Pkcs;
-    using Org.BouncyCastle.Security;
-    using X509Certificate = Org.BouncyCastle.X509.X509Certificate;
 
     public static class CertificateHelper
     {
+        // The private-key import on windows randomly seems failing, however according to the tests, the second time
+        // after a failure it usually works. The number below is just a "big enough" number randomly chosen for
+        // self-healing, but gives a limit to avoid endless try.
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs:21]: Do we need to combine the 2 fixes? I would rather keep them separate... unless we need to do this as part of BC removal, for some reason.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 2]
```diff
@@ -9,18 +9,21 @@ namespace Microsoft.Azure.Devices.Edge.Util
...
+        const int MaxCertImportRetryCount = 10;
+
+        private static Oid oidRsaEncryption = Oid.FromFriendlyName("RSA", OidGroup.All);
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs:24]: Nit: remove private qualifier

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 3]
```diff
@@ -462,5 +422,126 @@ static Option<string> GetCommonNameFromSubject(string subject)
...
+        {
+            int retryCount = 0;
+            while (retryCount++ < MaxCertImportRetryCount)
+            {
+                string pkcs8Label = "PRIVATE KEY";
+                string rsaLabel = "RSA PRIVATE KEY";
+                string ecLabel = "EC PRIVATE KEY";
+
+                var keyAlgorithm = certificate.GetKeyAlgorithm();
+
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs:436]: Nit: remove whitespace..

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 4]
```diff
@@ -462,5 +422,126 @@ static Option<string> GetCommonNameFromSubject(string subject)
...
+                }
+
+                if (result == null)
+                {
+                    throw new InvalidOperationException($"Cannot use certificate, not supported key algorithm: ${keyAlgorithm}");
+                }
+
+                try
+                {
+                    // On Windows the certificate in 'result' gives an error when used with kestrel: "No credentials are available in the security"
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs:488]: If this is Windows only, maybe we can keep it to release/1.1 only?

File: edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs [hunk 5]
```diff
@@ -462,5 +422,126 @@ static Option<string> GetCommonNameFromSubject(string subject)
 
             return commonName;
         }
+
+        static X509Certificate2 AttachPrivateKey(X509Certificate2 certificate, string pemEncodedKey)
+        {
+            int retryCount = 0;
+            while (retryCount++ < MaxCertImportRetryCount)
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Util/CertificateHelper.cs:429]: If this code is Windows specific.. then maybe we don't need it in master?

**Comment by varunpuranik** [comment]
> What missing are:
> 
> * error handling
> * Import apparently does not like PKCS-8

Is this still applicable? I do see PKCS-8 support. 
Otherwise changes look good.

**Review by varunpuranik — DISMISSED** [review]

**Comment by darobs** [comment]
#6051 should fix the E2E checkin build failure.

**Review by varunpuranik — APPROVED** [review]

**Pull request merged**